source/change-streams/change-streams.rst

```diff
@@ -405,28 +454,45 @@ If neither ``startAtOperationTime`` nor ``resumeAfter`` are specified, and the m
 resumeAfter
 ^^^^^^^^^^^
 
-``resumeAfter`` is used to resume a changeStream that has been stopped to ensure that only changes starting with the log entry immediately *after* the provided token will be returned. If the resume token specified does not exist, the server will return an error. 
+``resumeAfter`` is used to resume a change stream that has been stopped to ensure that only changes starting with the log entry immediately *after* the provided token will be returned. If the resume token specified does not exist, the server will return an error.
+
+ChangeStream::resumeToken
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The locally cached ``ChangeStream::resumeToken`` MUST be updated by the following rules.
+
+Upon creating a ``ChangeStream``, if ``resumeAfter`` is set in ``ChangeStreamOptions``, then ``ChangeStream::resumeToken`` MUST be set to the value of ``resumeAfter``.
+
+Upon sending the initial ``aggregate`` or a ``getMore`` command, if the response includes ``resumeTokens``, the following rules apply:
+- If the response has an empty batch (no documents) then ``ChangeStream::resumeToken`` MUST be set to ``resumeTokens[0]`` (the only element in ``resumeTokens``).
+- Otherwise, when document ``i`` of the current batch is returned, ``ChangeStream::resumeToken`` MUST be set to ``resumeTokens[i]``.
```

Comment:
In https://github.com/mongodb/specifications/pull/413#discussion_r231948022, @gormanb proposed skipping the resume token of the last document in the batch in favor of the final token (i.e. `minPromisedResumeToken`).

Alternatively, that last `minPromisedResumeToken` can get picked up when the user iterates beyond that last document in the local batch, immediately before a `getMore` would be issued. In that case, there is likely no point where `getResumeToken()` would return the `minPromisedResumeToken`, as the next time the application has control would be after the `getMore` response is processed.

These two bullets under "response includes `resumeTokens`" do seem a bit inconsistent in that:

 * When the batch is empty, we immediately pick up on `resumeTokens[0]`.
 * When the batch isn't empty, we only update `ChangeStream::resumeToken` when advance the iterator.

In the empty batch case, `getResumeToken()` would immediately have a new value. That's not necessarily the case with the non-empty batch case, although perhaps that depends on the implementation details for the iterators. For instance, does the initial `watch()` method return a change stream iterator already positioned at the first document or must users advance it to the first position?

Reply:
> In the empty batch case, getResumeToken() would immediately have a new value. That's not necessarily the case with the non-empty batch case, although perhaps that depends on the implementation details for the iterators. For instance, does the initial watch() method return a change stream iterator already positioned at the first document or must users advance it to the first position?

Maybe I'm missing your point. But in the non-empty case, we can only update the cached resume token until after we've returned the document. This shouldn't depend on the implementation details of the iterator.

Reply:
This might explain what I was trying to get across:

#### Example with empty batches

1. `Collection::watch()` is invoked and `aggregate` returns an empty batch with a single `minPromisedResumeToken`. The ChangeStream iterator is returned.
1. `ChangeStream::getResumeToken()` returns that `minPromisedResumeToken`, since it was picked up immediately from the `aggregate` response.
1. Accessing the current position of the ChangeStream iterator returns a disengaged optional (i.e. no document).
1. Advancing the ChangeStream iterator issues a `getMore` command, which still returns an empty batch with a single `minPromisedResumeToken`.
1. `ChangeStream::getResumeToken()` returns the `minPromisedResumeToken` from the last `getMore`.
1. Accessing the current position of the ChangeStream iterator still returns a disengaged optional (i.e. no document).

In this case of empty batches, `ChangeStream::getResumeToken()` will return the same `minPromisedResumeToken` no matter whether it is called before or after access of the ChangeStream iterator's current document. Advancing the iterator and invoking a `getMore` will certainly see the resume token updated.

#### Example with non-empty batches

1. `Collection::watch()` is invoked and `aggregate` returns a batch with two documents. The `resumeTokens` array contains the first document's resume token and a trailing `minPromisedResumeToken`. The ChangeStream iterator is returned.
1. `ChangeStream::getResumeToken()` returns nothing, since we've yet to access/return a document. (note: this seems like a subtle behavioral change)
1. Accessing the current position of the ChangeStream iterator returns the _first_ document in the batch. When this document is accessed, the internal `ChangeStream.resumeToken` property is updated.
1. `ChangeStream::getResumeToken()` now returns the resume token for the _first_ document.
1. The ChangeStream iterator is advanced.
1. Accessing the current position of the ChangeStream iterator returns the _second_ document in the batch. When this document is accessed, the internal `ChangeStream.resumeToken` property is updated. (note: I believe  @gormanb's proposal would see the `resumeToken` updated to `minPromisedResumeToken` at this point)
1. `ChangeStream::getResumeToken()` now returns the resume token for the _second_ document.
1. The ChangeStream iterator is advanced. Before the `getMore` is issued, I believe we'd want to update `resumeToken` to `minPromisedResumeToken`; however, this doesn't appear to be outlined in the spec as written (unless I'm missing something). `getMore` returns a new batch containing a few documents.
1. `ChangeStream::getResumeToken()` returns the `minPromisedResumeToken` from the previous batch, since we've yet to access/return a document from the second batch returned by `getMore`.

Is this sequence of events (for both cases) accurate?

Reply:
> Example with empty batches

This is accurate.

> Example with non-empty batches
> [...]
> 2. ChangeStream::getResumeToken() returns nothing [...]

Yes, if this is the very first response. If there was a previous response (empty or non-empty) it would return the last cached resume token.

> 6. Accessing the current position of the ChangeStream iterator returns the second document in the batch. When this document is accessed, the internal ChangeStream.resumeToken property is updated. (note: I believe @gormanb's proposal would see the resumeToken updated to minPromisedResumeToken at this point)
> 7. ChangeStream::getResumeToken() now returns the resume token for the second document.

The current proposal written here (which will likely change) assumes that `resumeTokens` would have length 2 in this case. `resumeTokens[0]` is the resume token for the first document, and `resumeTokens[1]` is the `minPromisedResumeToken`. As soon as the second document is iterated, the cached resume token becomes the value of `minPromisedResumeToken`.

Reply:
Per our offline conversation, I think this can be resolved.

@kevinAlbs explained that, for non-empty batches, the server will likely replace the last resume token in the `resumeTokens` array with the `minPromisedResumeToken` instead of the last document's actual resume token, since it is both functionally equivalent and more efficient. Therefore, the size of `resumeTokens` is either one (in the case of an empty batch) or exactly the same size as the number of change documents. We agreed that the spec would do well to have a FAQ entry added to document this, but we don't need to describe this behavior in detail within this section of the spec.